components/kyma-environment-broker/internal/runtime/components_provider.go

```diff
@@ -0,0 +1,346 @@
+		}
+	}
+
+	return requiredComponents, nil
+}
+
+func (p *defaultRequiredComponentsProvider) getComponentsYamlURL(kymaVersion string) string {
+	if p.isOnDemandRelease(kymaVersion) {
+		return fmt.Sprintf(onDemandComponentsURLFormat, kymaVersion)
+	}
+	return fmt.Sprintf(releaseComponentsURLFormat, kymaVersion)
+}
+
+// isOnDemandRelease returns true if the version is recognized as on-demand.
+//
+// Detection rules:
+//   For pull requests: PR-<number>
+//   For changes to the main branch: main-<commit_sha>
+//
+// source: https://github.com/kyma-project/test-infra/blob/main/docs/prow/prow-architecture.md#generate-development-artifacts
+func (p *defaultRequiredComponentsProvider) isOnDemandRelease(version string) bool {
+	isOnDemandVersion := strings.HasPrefix(version, "PR-") ||
+		strings.HasPrefix(version, "main-")
+	return isOnDemandVersion
+}
+
+func (p *defaultRequiredComponentsProvider) checkStatusCode(resp *http.Response) error {
+	if resp.StatusCode == http.StatusOK {
+		return nil
+	}
+
+	// limited buff to ready only ~4kb, so big response will not blowup our component
+	body, err := io.ReadAll(io.LimitReader(resp.Body, 4096))
+	if err != nil {
+		body = []byte(fmt.Sprintf("cannot read body, got error: %s", err))
+	}
+	msg := fmt.Sprintf("while checking response status code for Kyma components list: "+
+		"got unexpected status code, want %d, got %d, url: %s, body: %s",
+		http.StatusOK, resp.StatusCode, resp.Request.URL.String(), body)
+
+	switch {
+	case resp.StatusCode == http.StatusRequestTimeout:
+		return kebError.NewTemporaryError(msg)
+	case resp.StatusCode >= http.StatusInternalServerError:
+		return kebError.NewTemporaryError(msg)
+	default:
+		return errors.New(msg)
+	}
+}
+
+func (p *defaultAdditionalComponentsProvider) AdditionalComponents(kymaVersion internal.RuntimeVersionData, plan string) ([]KymaComponent, error) {
+	if kymaVersion.MajorVersion >= 2 {
+		defaults, err := p.fetchDefaultAdditionalComponents()
+		if err != nil {
+			return nil, fmt.Errorf("while fetching default additional components: %w", err)
+		}
+
+		custom, err := p.fetchAdditionalComponentsForGivenVersionAndPlan(kymaVersion, plan)
+		if err != nil {
+			return nil, fmt.Errorf("while fetching additional components for version %s and plan %s: %w",
+				kymaVersion.Version, plan, err)
+		}
+
+		if custom == nil {
+			return defaults, nil
+		}
+
+		return custom, nil
+	}
+	return nil, errors.New("unsupported Kyma version")
+
+}
+
+func (p *defaultAdditionalComponentsProvider) fetchDefaultAdditionalComponents() ([]KymaComponent, error) {
+	yamlContents, err := p.readYaml(p.defaultAdditionalComponentsYamlPath)
+	if err != nil {
+		return nil, err
+	}
+	return p.getComponentsFromYaml(yamlContents)
+}
+
+func (p *defaultAdditionalComponentsProvider) fetchAdditionalComponentsForGivenVersionAndPlan(kymaVersion internal.RuntimeVersionData, plan string) ([]KymaComponent, error) {
+	if plan == "" {
+		return nil, nil
```

Comment:
nit: how about logging this case?

Reply:
I like the idea, however we will need to pass a logger to the component provider or place it in the input builder factory. We will need to switch component providers sometime in the future, let's try to add logger then.